LCD position in current units

rdhoggattjr opened the pull request:
### Requirements

* Filling out this template is required. Pull Requests without a clear description may be closed at the maintainers' discretion.

### Description

Adds an option if 'INCH_MODE_SUPPORT' is enabled to show XYZ Coordinate's in inches on the status screen when G20 is active.

### Benefits

INCH_MODE_SUPPORT lets you use inches but still display's the current position in mm. This changes that.

### Configurations

Sorry, I've never contributed on Github before so I don't know if this is right but I'll put a basic Configuration.h with the display I
 have turned on.
[Configuration.txt](https://github.com/MarlinFirmware/Marlin/files/5541853/Configuration.txt)

### Related Issues

<!-- Whether this fixes a bug or fulfills a feature request, please list any related Issues here. -->

rdhoggattjr committed:
Merge pull request #1 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

rdhoggattjr committed:
INFO_DISPLAY_INCHES

Display inches on status screen when using G20

rdhoggattjr committed:
Fix formatting

rdhoggattjr committed:
STATUS_DISPLAY_INCHES

Display XYZ Coordinate's in Inches when G20 is active. Also turn on 'LCD_DECIMAL_SMALL_XY' and 'XYZ_NO_FRAME' as they are needed for STATUS_DISPLAY_INCHES.

rdhoggattjr committed:
Re-Push STATUS_DISPLAY_INCHES

Push the rest of the changes that Github Desktop missed for some reason.

rdhoggattjr committed:
Update Configuration.h

rdhoggattjr committed:
Remove white space and fix check

rdhoggattjr commented:
I have a possibly dumb question, but if I rebase and squish all the commits will it update the PR? And if so will it mess up the PR in some way?

Sorry for the questions but I'm new to Git/Github and still trying to learn.

sjasonsmith commented:
@rdhoggattjr if you rebase you have to force-push to your branch to update the PR. It causes the review to lose some continuity because you can't see the changes made between commits if you rebase. I believe comments made on the code also don't track properly across a rebase.

I believe that when thinkyhead merges PRs they are usually done a squash merge, which is going to condense this all down to one commit. You don't need to worry about there being a large number of commits since that will happen anyway.

sjasonsmith commented on Marlin/src/lcd/dogm/status_screen_DOGM.cpp:
```diff
@@ -371,9 +380,22 @@ FORCE_INLINE void _draw_centered_temp(const int16_t temp, const uint8_t tx, cons
-  lcd_put_wchar(X_LABEL_POS + offs, XYZ_BASELINE, axis_codes[axis]);
-  lcd_moveto(X_VALUE_POS + offs, XYZ_BASELINE);
+  #if ENABLED(STATUS_DISPLAY_INCHES)
+    if (parser.linear_unit_factor == 25.4f) {
```
You need a better way to check for inches than repeating this `25.4f` constant in multiple places.
The parser already has `set_input_linear_units`, you should probably create a `get_input_linear_units` that returns the `LinearUnit` enum.

rdhoggattjr replied:
Thank you for your input. I'm a self taught programmer, if you can even call me that.

I think what your trying to say is I need to make some sort of function that checks and call that function instead

Basically it checks and sets a bool so as to not mess with the float but once a cycle?

sjasonsmith replied:
It is more that I don't want to see 25.4f defined in multiple places. If it is in only one place, it is impossible for them to disagree.

Since there it already a `set_input_linear_units` is seems proper to add a function next to it to get the current state.

rdhoggattjr replied:
I tried adding a function `get_input_linear_units` but had an issue with variable types, so I went with a different approach for the time being. It built and tested fine on my hardware though.

I'll keep working on a separate function for the boolean if need be.

rdhoggattjr replied:
@sjasonsmith I was able to get a separate function going, but to me 21edaeb has easier to follow code. I can push the function code if you would like to see it.

sjasonsmith commented:
It hurts me a little to see the expansion of "inch" capability, and I know this PR has been a little controversial in some of the forums for that reason. That said, if using things alongside existing inch-based CNC machinery I can appreciate the desire for some consistency. I myself have to frequently switch between mm and inches depending on whether I am using a 3D printer or a CNC mill.

rdhoggattjr commented:
Yeah, I'm using Marlin with a CNC converted Mill that uses inches already.

rdhoggattjr committed:
add imperial_units boolean

Now checks imperial_units boolean for linear units in use.

rdhoggattjr commented:
I'm thinking I may need to put in some check for the LCD being used as this only works with certain controllers?

How does testing normally work for changes when you don't have access to every setup? Or do I limit it to known compatible hardware at the moment and leave it for someone with said hardware to port it?

rdhoggattjr committed:
Update Configuration.h

rdhoggattjr committed:
Merge pull request #3 from rdhoggattjr/bugfix-2.0.x-temp

add imperial_units boolean

rdhoggattjr committed:
added INCH_MANUAL_MOVEMENT

added code for INCH_MANUAL_MOVEMENT testing

rdhoggattjr committed:
Adjust conditions for imperial_units boolean

use imperial_units boolean for either STATUS_DISPLAY_INCHES or INCH_MANUAL_MOVEMENT

rdhoggattjr committed:
small fixes

made small changes

rdhoggattjr committed:
fix #include check

rdhoggattjr committed:
Merge pull request #4 from rdhoggattjr/bugfix-2.0.x-temp

fix #include check

rdhoggattjr committed:
More Small Fixes

This looks ready to test on the machine.

thinkyhead commented:
Since the idea is to make inch support more complete across the UI, the UI should simply reflect the current units as set by `M20` / `M21` without need for a new option. The code can just use the `LINEAR_UNIT` macro in `parser.h` to display the correct value, converted from mm. And it can test whether inches mode is enabled with the addition of…
```cpp
static inline bool using_inch_units() { return mm_to_linear_unit(1.0f) != 1.0f; }
```
…to decide whether to display `mm` for millimeters or `in` (or possibly `"`) for inches, where appropriate. Marlin is already set up so that the output of `M503` and other codes that report positions and lengths will automatically output inches when in inches mode.

thinkyhead committed:
Helper method

rdhoggattjr committed:
Merged STATUS_DISPLAY_INCHES into INCH_MODE_SUPPORT

Now enabled if INCH_MODE_SUPPORT is enabled and not an option

rdhoggattjr commented:
@thinkyhead 5aff43e is what I think you were getting at? If so I'll test it on my test bed tonight.

Thank you for your input also.

rdhoggattjr commented:
Sorry for all the questions, I just want to make sure I'm doing things right. 

I also have code that changes the move axis menu to inch when G20 is active. My question is would it be better to start a new pull request for that or commiting it here since this seems to be changing from an option to working towards completing one?

rdhoggattjr committed:
Merge branch 'bugfix-2.0.x-inch-manual-movement' into bugfix-2.0.x-temp

rdhoggattjr committed:
Merge pull request #5 from rdhoggattjr/bugfix-2.0.x-temp

Bugfix 2.0.x temp

rdhoggattjr committed:
Convert to INCH_MODE_SUPPORT and using_inch_units()

Tied INCH_MANUAL_MOVEMENT into INCH_MODE_SUPPORT and converted unit checks to using_inch_units()

rdhoggattjr committed:
Merge pull request #6 from rdhoggattjr/bugfix-2.0.x-inch-manual-movement

Bugfix 2.0.x inch manual movement

rdhoggattjr commented:
@thinkyhead Some how I merged the branch that I was using for the manual movement in inches into the pull request anyways. Sorry! If you don't want it in there it can be deleted starting with 9e26439. I pulled the rest just to make it complete since it was there already.

I was going to do it but can't seem to find where.

rdhoggattjr committed:
Update to use TERN

Condensed IF ENABLED statements to use TERN for cleaner code.

rdhoggattjr committed:
Remove whitespace

thinkyhead commented:
When it comes to PRs, you can tweak your command history in your working copy all you like then use `git push -f` to update the PR to your current state. But, if this PR's "Files" changes all look correct then it can go forward from here.

rdhoggattjr commented:
> When it comes to PRs, you can tweak your command history in your working copy all you like then use `git push -f` to update the PR to your current state. But, if now this PR's "Files" changes all look correct then it can go forward from here.

@thinkyhead Everything seems to be looking good. Now I'm just going though the code to try and make it look cleaner, then it should be good for hopefully a final review from someone.

thinkyhead committed:
Simplify

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20145

thinkyhead committed:
Better macro

thinkyhead committed:
Include parser

rdhoggattjr commented:
@thinkyhead I wish I was more knowledgeable in coding. You make it look easy. If you don't mind me asking, what happens now?

sjasonsmith commented:
@rdhoggattjr did you test with all the recent changes thinyhead added?

I believe his changes make it so that it will switch as you switch units through gcode. It would be great if you can test that ability to switch back and forth.

rdhoggattjr commented:
> @rdhoggattjr did you test with all the recent changes thinyhead added?
> 
> I believe his changes make it so that it will switch as you switch units through gcode. It would be great if you can test that ability to switch back and forth.

@sjasonsmith It works great on my test bed. I have a mega2560 with ramps and full discount graphics sitting next to my chair that I use to test as I'm coding.

The only thing that I seen was that M114 still reports mm all the time, but I wasn't going to mess with it because of  Octoprint or others that might use it and could possibly cause issues.

thinkyhead committed:
M114 tweak
